content/posts/numpy/numpy-rng/index.md

```diff
@@ -0,0 +1,110 @@
+---
+title: "Good practices with numpy random number generators"
+date: 2023-08-31T23:22:46+02:00
+draft: false
+description: "
+Good practices with numpy random number generators
+"
+tags: ["tutorials", numpy, rng]
+displayInList: true
+author: ["Albert Thomas"]
+---
+
+Unless you are working on a problem where you can afford a true Random Number Generator (RNG), which is basically never for most of us, implementing something random means relying on a pseudo Random Number Generator. I want to share here what I have learned about good practices with pseudo RNGs and especially the ones available in [numpy](https://numpy.org/). I assume a certain knowledge of numpy and that numpy 1.17 or greater is used. The reason for this is that great new features were introduced in the [random](https://numpy.org/doc/1.18/reference/random/index.html) module of version 1.17. As `numpy` is usually imported as `np`, I will sometimes use `np` instead of `numpy`. Finally, as I will not talk about true RNGs, RNG will always mean pseudo RNG in the rest of this blog post.
+
+### The main messages
+1. Avoid using the global numpy RNG. This means that you should avoid using [`np.random.seed`](https://numpy.org/doc/1.18/reference/random/generated/numpy.random.seed.html) and `np.random.*` functions, such as `np.random.random`, to generate random values.
+2. Create a new RNG and pass it around using the [`np.random.default_rng`](https://numpy.org/doc/1.18/reference/random/generator.html#numpy.random.default_rng) function.
+3. Be careful with parallel computations and rely on [numpy strategies for reproducible parallel number generation](https://numpy.org/doc/1.18/reference/random/parallel.html).
+
+Note that with numpy <1.17 the way to create a new RNG is to use [`np.random.RandomState`](https://numpy.org/doc/1.18/reference/random/legacy.html#numpy.random.RandomState) which is based on the popular Mersenne Twister 19937 algorithm. This is also how the global numpy RNG is created. It is still possible to use this function in versions higher than 1.17 but it is now recommended to use `default_rng` which returns an instance of the statistically better [PCG64](https://www.pcg-random.org) RNG.
```

Comment:
It might be worth noting a slight nuance (either here or in some advance section or footnote). `np.random.RandomState` is going to be around and there is no plan to mark it at legacy either. This is pretty much the recommended RNG to use for testing as there are still the same guarantee as before in terms of byte stream stability. Which the new infra relaxes a bit.

Reply:
> This is pretty much the recommended RNG to use for testing as there are still the same guarantee as before in terms of byte stream stability

you mean for existing projects so that tests won't break?

Reply:
Yes. But also if you write new tests, the official recommended way is to use `RandomState`. I agree that it's not clear from their doc, but maintainers of NumPy told me that a few times.

Reply:
is it that Mersenne Twister has better properties for testing?

Reply:
No the reason is that because they still provide strong guarantees of stability with `RandomState`, the tests are less likely to break. But because of known statistical issues, it's not recommended to use these for anything else than production. Bugs are known and cannot be fixed due to this same guarantee.

TBH, I personally only use `Generator` for all my code for simplicity. It's really not like this is going to break all the time.

Reply:
Ok got it. I added a sentence, but do we want to discourage this kind of practice (even for tests) or at least not encourage it?